src/lsdb/loaders/hats/read_hats.py

```diff
@@ -69,23 +60,169 @@ def read_hats(
+        dask_df_pixel_map = {}
+    return AssociationCatalog(dask_df, dask_df_pixel_map, hc_catalog)
 
-    catalog_type_to_use = _get_dataset_class_from_catalog_info(path)
 
-    if catalog_type is not None:
-        catalog_type_to_use = catalog_type
+def _load_margin_catalog(hc_catalog, config):
+    """Load a catalog from the configuration specified when the loader was created
 
-    loader = get_loader_for_type(catalog_type_to_use, path, config)
-    return loader.load_catalog()
+    Returns:
+        Catalog object with data from the source given at loader initialization
+    """
+    if config.search_filter:
+        filtered_catalog = config.search_filter.filter_hc_catalog(hc_catalog)
+        hc_catalog = hc.catalog.MarginCatalog(
+            filtered_catalog.catalog_info,
+            filtered_catalog.pixel_tree,
+            catalog_path=hc_catalog.catalog_path,
+            schema=filtered_catalog.schema,
+        )
+    dask_df, dask_df_pixel_map = _load_dask_df_and_map(hc_catalog, config)
+    margin = MarginCatalog(dask_df, dask_df_pixel_map, hc_catalog)
+    if config.search_filter is not None:
+        margin = margin.search(config.search_filter)
+    return margin
+
+
+def _load_object_catalog(hc_catalog, config):
+    """Load a catalog from the configuration specified when the loader was created
 
+    Returns:
+        Catalog object with data from the source given at loader initialization
+    """
+    if config.search_filter:
+        filtered_catalog = config.search_filter.filter_hc_catalog(hc_catalog)
+        if len(filtered_catalog.get_healpix_pixels()) == 0:
+            raise ValueError("The selected sky region has no coverage")
+        hc_catalog = hc.catalog.Catalog(
+            filtered_catalog.catalog_info,
+            filtered_catalog.pixel_tree,
+            catalog_path=hc_catalog.catalog_path,
+            moc=filtered_catalog.moc,
+            schema=filtered_catalog.schema,
+        )
 
-def _get_dataset_class_from_catalog_info(base_catalog_path: str | Path | UPath) -> Type[Dataset]:
-    base_catalog_dir = hc.io.file_io.get_upath(base_catalog_path)
-    catalog_info = TableProperties.read_from_dir(base_catalog_dir)
-    catalog_type = catalog_info.catalog_type
-    if catalog_type not in dataset_class_for_catalog_type:
-        raise NotImplementedError(f"Cannot load catalog of type {catalog_type}")
-    return dataset_class_for_catalog_type[catalog_type]
+    dask_df, dask_df_pixel_map = _load_dask_df_and_map(hc_catalog, config)
+    catalog = Catalog(dask_df, dask_df_pixel_map, hc_catalog)
+    if config.search_filter is not None:
+        catalog = catalog.search(config.search_filter)
+    if isinstance(config.margin_cache, MarginCatalog):
+        catalog.margin = config.margin_cache
+        if config.search_filter is not None:
+            # pylint: disable=protected-access
+            catalog.margin = catalog.margin.search(config.search_filter)
```

Comment:
When the margin is provided as an object we're applying the spatial filters but the other parquet arguments (e.g. columns, filters) are unfortunately being ignored. If the margin cache is provided as a path (second branch of this if) the parquet arguments are applied because they are pushed down to the `read_parquet_file_to_pandas` call. Could we just assume that if the margin is provided as an object, it has already been preprocessed?

// This is not related to the refactoring and can definitely be handled in a separate issue!

Reply:
We had discussed last week to just remove the option to provide the margin as an object, and only as a path. I think it addresses enough of the mismatch concerns, and is also likely the more common route for users.